src/twisted/conch/ssh/transport.py

```diff
@@ -516,6 +515,99 @@ class SSHTransportBase(protocol.Protocol):
+        Then call addSupportedKexes to populate the list.
+
+        @type kexes: L{list}
+        @param kexes: A L{list} of kex name strings
+        """
+        # Clear out the currently supported key exchanges.
+        _kex._kexAlgorithms.clear()
+
+        # Set the new exchanges.
+        self.addSupportedKexes(kexes)
+
+
+    def addSupportedKexes(self, kexes):
+        """
+        Append the list of supported key exchanges.
+
+        @type kexes: L{list}
+        @param kexes: A L{list} of kex name strings
+        """
+        for k in kexes:
+            if k.startswith(b'ecdh') and kexes[k] is None:
+                size = int(k[-3:])
+
+                # Add the new kex to the dictionary.
+                if size <= 256:
+                    _kex._kexAlgorithms[k] = _kex._ECDH256()
+                elif size <= 384:
+                    _kex._kexAlgorithms[k] = _kex._ECDH384()
+                else:
+                    _kex._kexAlgorithms[k] = _kex._ECDH512()
+            elif kexes[k] is None:
+                _kex._kexAlgorithms[k] = _kex._baseKexAlgorithms[k]
+            else:
+                _kex._kexAlgorithms[k] = kexes[k]
+
+        # Reload
+        del self.supportedKeyExchanges[:]
+        for kex in _kex.getSupportedKeyExchanges():
+            self.supportedKeyExchanges.append(kex)
+
+
+    def setSupportedPublicKeys(self, publicKeys):
+        """
+         Clear out the list of supported public keys
+         Then call addSupportePublicKeys to populate the list.
+
+         @type publicKeys: L{dict}
+         @param publicKeys: A L{dict} with the name of the public key as the key
+            and a value that is a L{tuple} consisting of an instance of the key method,
+            a L{str} that is the corresponding sec name.
+        """
+        del self.supportedPublicKeys[:]
+        self.addSupportePublicKeys(publicKeys)
+
+
+    def addSupportePublicKeys(self, publicKeys, sort=True):
+        """
+        Populate the supported public key list.
+
+        @type publicKeys: L{dict}
+        @param publicKeys: A L{dict} with the name of the public key as the key
+            and a value that is a L{tuple} consisting of an instance of the key method,
+            a L{str} that is the corresponding sec name.
+        """
+        for k in publicKeys:
+            if k not in self.supportedPublicKeys:
+                self.supportedPublicKeys.append(k)
+
+            if k.startswith(b'ecdsa'):
+                curve = publicKeys[k]
+                keys._curveTable[k] = curve[0]
+
+                keys._secToNist[curve[1]] = k
+
+        # Sort the keys so the strongest ones are preferred?
+        if sort:
+            # Reverse sort to bring the stronger keys up first.
+            self.supportedPublicKeys.sort(reverse=True)
+
+            # If 'ssh-rsa' or 'ssh-dss' are present, remove and add
+            # to make sure they're in the right order
+            if b'ssh-rsa' in self.supportedPublicKeys:
```

Comment:
Neither am I (see above).

Sorting public keys is a problem because there's no way to designate which one is preferred like there is with key exchanges.  But like kexes they should be ordered so the most preferred ones are used.

The default curves can put the strongest key first by a reverse list sort, and then `ssh-rsa` and `ssh-dsa` can be appended on after.  This is a good first step, but there's room for improvement.  And if `sort` is False then the order will be whatever is passed into `addSupportedKexes()`

I think the link you provided would work if in the future Twisted supported adding keys via a cli argument. Plus the API provided allows for the addition of crypto backends to be included for ciphers that aren't supported by Cryptography.